protocol/messenger_profile_showcase.go

```diff
@@ -107,6 +109,48 @@ func (m *Messenger) validateCollectiblesOwnership(accounts []*identity.ProfileSh
 	return nil
 }
 
+func (m *Messenger) validateCommunitiesMembership(communities []*identity.ProfileShowcaseCommunity, contactPubKey *ecdsa.PublicKey) ([]*identity.ProfileShowcaseCommunity, error) {
+	validatedCommunities := []*identity.ProfileShowcaseCommunity{}
+
+	for _, communityEntry := range communities {
+		community, err := m.FetchCommunity(&FetchCommunityRequest{
+			CommunityKey:    communityEntry.CommunityID,
+			Shard:           nil,
+			TryDatabase:     true,
+			WaitForResponse: true,
+		})
+		if err != nil {
+			m.logger.Warn("failed to fetch community for profile entry ", zap.Error(err))
+		}
+
+		if community != nil && community.Encrypted() {
+			grant, err := community.VerifyGrantSignature(communityEntry.Grant)
+			if err != nil {
+				m.logger.Warn("failed to verify grant signature ", zap.Error(err))
+				communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusNotAMember
+			} else {
+				if grant != nil && bytes.Equal(grant.MemberId, crypto.CompressPubkey(contactPubKey)) {
+					communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusProvenMember
+				} else { // Show as not a member if membership can't be proven
+					communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusNotAMember
+				}
+			}
+		} else if community != nil {
+			// Use member list as a proof for unecrypted communities
+			if community.HasMember(contactPubKey) {
+				communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusProvenMember
+			} else {
+				communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusNotAMember
+			}
+		} else {
+			communityEntry.MembershipStatus = identity.ProfileShowcaseMembershipStatusUnproven
+		}
```

Comment:
The logic seems very complex, there seems to be a lot of branches to go through, could be possibly have simplified logic here?

Reply:
Right, i refactor it to a function